icefoganalytics/internal-data-portal
# Part 2: Implement Improvements Developed in Meeting On 2024-03-27

Fixes https://github.com/icefoganalytics/internal-data-portal/issues/92

# Context

## Problem Statement

Amalgamating data from various sources in a consistent way comes with several caveats in regards to data normalization and storage. 

Currently the data portal is doing a full import of an external dataset and this may not scale well, or add much value to end users.

End users will use the platform to gain a quick preview of the dataset, and get an example of data quality. If they want more data, or a real-time interaction with the dataset, they will interact with the source dataset directly.

## Summary

- [x] Give end users the ability to refresh the current dataset preview.
- [x] Provide ability to add local datasets that aren’t previews.


# Implementation

Set api base url from window.location.origin in production to help direct links break out of vue-router control.
Add datasets/:slugOrId/refresh endpoint and controller to refresh dataset from remote source.
Central dataset policy "show" logic. Show now supports a limited and unlimited access variation.
Permit changing dataset owner via the dataset description "owner" panel update.
Prevent deletion of a user while they are the owner of a dataset as this make the dataset effectively invisible due to policy scoping logic. See https://icefoganalytics.slack.com/archives/C064Q5KA0R0/p1712527181956819
Allow importing and export dataset from CSV.

# Screenshots

New "refresh dataset" button that is accessible to all users that have "unlimited" view access to the dataset.
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/f79cb327-703a-4384-bd97-c9304382395a)

New Upload Data button on dataset description manage page.
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/40a41798-7767-4c4b-a9a5-58a0440ad203)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080 as a "data owner".
4. Begin creating a new dataset via the "Add dataset" button in the top right of the dashboard.
5. Call the dataset "QA-ing Dataset Integration Refresh".
6. Make yourself the data owner.
7. Create the dataset and get redirected to the dataset description manage page.
8. Add an API link via the button by the dataset name.
9. Use the following data to create the integration
    API link
    ```
    https://api.gov.yk.ca/directory/employees
    ```
    Header Key
    ```
    Ocp-Apim-Subscription-Key
    ```
    Header Value: as found in your .env file.
10. After creating the integration, set the parsing expressing to `employees` and "import" the dataset.
11. The UI should now show _only_ the "Manage API Link" button.
12. Go to the "Fields" tab and add some fields from the dataset such as:
    - `full_name`
    - `email`
    - `department`
    - `title`
    - `address`
13. Go to the "Visualize" tab.
14. Check that you can now see some data.
15. Click the "refresh" icon to refresh the dataset, it takes a while. It also fails pretty often, just try again.
16. Toggle the "allow download to CSV" flag.
17. Download the CSV somewhere convenient, you'll need it to QA the next part. Note that download operates on the full dataset, you can filter it if you want via search. It will take a while, and might fail for the same reason as the refresh. Just try again.
18. Once you have the CSV downloaded, go back to the all datasets table and begin creating a new dataset.
19. Call the dataset "QA-ing Dataset Upload".
20. Make yourself the data owner.
21. Create the dataset and get redirected to the dataset description manage page.
22. Click the new "Upload Data" button to open the upload modal.
23. Pick the export you created previously.
24. Go to the Fields tab and note that the fields are auto-generated.
25. Go to the Visualize tab and see that your table is now "full sized" and not limited by the external size limit.
26. Check that "refresh" works, it will be near instant because it only refreshes the UI view.
27. Enable CSV download.
28. Download the full dataset. It is very fast because it doesn't have to wait for the integration to refresh.
29. Filter the dataset with search.
30. Download the filtered dataset. Search will now match download, as view is against full data, not against a limited preview.
